contract/vault/src/lib.rs

```diff
@@ -0,0 +1,1505 @@
+
+    underlying_asset: FungibleAsset<BorrowAsset>,
+    /// configuration per market (market ID -> MarketConfig)
+    config: IterableMap<AccountId, MarketConfiguration>,
+
+    // TODO: decimal offset for virtual shares
+    /// Performance fee (as WAD fraction)
+    performance_fee: wad::WADFraction,
+    fee_recipient: AccountId,
+    skim_recipient: AccountId,
+    /// Last recorded total assets (for fee accrual)
+    last_total_assets: u128,
+
+    // Virtual offsets used only in conversions/previews to harden edge cases
+    virtual_shares: u128,
+    virtual_assets: u128,
+
+    /// Any pending change to the vault's cap, TODO: u256
+    pending_cap: IterableMap<AccountId, PendingValue<u128>>,
+    /// Any pending change to the vault's timelock
+    pending_timelock: Option<PendingValue<TimestampNs>>,
+    /// Any pending change to the vault's guardian
+    pending_guardian: Option<PendingValue<AccountId>>,
+    /// Current timelock duration for governance actions (ns)
+    timelock_ns: TimestampNs,
+
+    // Ordered list of market IDs for deposit allocation
+    supply_queue: Vector<AccountId>,
+    // Ordered list of market IDs for withdrawal prioritytr
+    withdraw_queue: Vector<AccountId>,
+
+    // vault's supplied principal per market (borrow-asset units)
+    market_supply: LookupMap<AccountId, u128>,
+
+    // underlying held by vault
+    idle_balance: u128,
+    op_state: OpState,
+    next_op_id: u64,
+
+    // Storage usage
+    storage_usage_supply: u64,
+    storage_usage_role: u64,
+
+    // Pending withdrawals queue (vault-level, FIFO by id)
+    pending_withdrawals: IterableMap<u64, PendingWithdrawal>,
+    next_withdraw_id: u64,
+    next_withdraw_to_execute: u64,
+}
+
+#[near]
+impl Contract {
+    #[allow(clippy::unwrap_used, reason = "Infallible")]
+    #[allow(clippy::too_many_arguments, reason = "Constructor")]
+    #[init]
+    /// Initializes a new vault.
+    /// - `owner_id`: account that controls Owner-only actions.
+    /// - `curator_id`: manages markets and is also granted the Allocator role.
+    /// - `guardian_id`: can revoke pending governance actions.
+    /// - `underlying_token_id`: NEP-141 underlying asset managed by the vault.
+    /// - `initial_timelock_sec`: governance timelock in seconds.
+    /// - `fee_recipient`: account to receive performance fees.
+    /// - `skim_recipient`: account to receive skimmed tokens.
+    /// - `name`/`symbol`/`decimals`: metadata for the share token.
+    pub fn new(configuration: VaultConfiguration) -> Self {
+        let VaultConfiguration {
+            owner,
+            curator,
+            guardian,
+            underlying_token,
+            initial_timelock_sec,
+            fee_recipient,
+            skim_recipient,
+            name,
+            symbol,
+            decimals,
+            mode,
+        } = configuration;
+
+        let timelock_ns = u64::from(initial_timelock_sec) * 1_000_000_000;
+        assert!(
```

Comment:
```suggestion
        require!(
```

should include fewer debug symbols